tests/test_backend_failure.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python3
+
+import logging
+import argparse
+
+from os import getcwd, chdir
+
+from test_fixtures import with_aktualizr, with_uptane_backend, KeyStore, with_secondary, with_path,\
+    DownloadInterruptionHandler, MalformedJsonHandler, with_director, with_imagerepo, InstallManager,\
+    with_install_manager, with_images, MalformedImageHandler, with_customrepo, SlowRetrievalHandler, \
+    RedirectHandler
+
+
+logger = logging.getLogger(__file__)
+
+"""
+Verifies whether aktualizr is updatable after director metadata download failure
+with follow-up successful metadata download.
+
+Currently, it's tested against two types of metadata download/parsing failure:
+    - download interruption - metadata file download is interrupted once|four times, after that it's successful
+    - malformed json - metadata file download is successful but the received metadata are malformed.
+      It happens at once| four time, after that it's successful
+
+
+Note: Aktualizr doesn't send any installation report in manifest in case of metadata download failure
+https://saeljira.it.here.com/browse/OTA-3730
+"""
+@with_uptane_backend(start_generic_server=True, port=8888)
+@with_path(paths=['/1.root.json', '/root.json', '/targets.json'])
+@with_director(handlers=[
+                            DownloadInterruptionHandler(number_of_failures=1),
+                            MalformedJsonHandler(number_of_failures=1),
+                            DownloadInterruptionHandler(number_of_failures=3),
+                            #MalformedJsonHandler(number_of_failures=3)
+                        ])
+@with_aktualizr(start=False, run_mode='full')
+@with_install_manager()
+def test_backend_failure_sanity_director_update_after_metadata_download_failure(install_mngr, director,
+                                                                                aktualizr, **kwargs):
+    with aktualizr:
+        director.wait_for_install()
+        # we have to stop director before terminating aktualizr since the later doesn't support graceful shutdown
+        # (doesn't handle any signal (SIGTERM, SIGKILL, etc) what leads to receiving broken requests at director
+        director.stop()
+    return director.is_install_successful() and install_mngr.are_images_installed()
+
+
+"""
+Verifies whether aktualizr is updatable after image metadata download failure
+with follow-up successful metadata download.
+
+Currently, it's tested against two types of metadata download/parsing failure:
+    - download interruption - metadata file download is interrupted once|four times, after that it's successful
+    - malformed json - metadata file download is successful but the received metadata are malformed.
+      It happens at once| four time, after that it's successful
```

Comment:
Looks like three times now; any significance to 3 vs 4?

Reply:
yeah, just removed.

Reply:
I'm confused, I still see the discrepancy here and elsewhere.

Reply:
It should be fine now, the comment was outdated.
No, there is no significance if it's 3,4 or even 5 :). There two use cases here:
1) download succeeds withing a single uptane cycle thanks to download retries, a number of retires is set to 3 at the moment (hardcoded in aktualizr)

2) download fails withing a single uptane cycle even after 3 attempts, aktualizr exits current uptane cycle, starts a new one and try to download the given metadata file again which succeeds.

So, those two DownloadInterruptionHandler(number_of_failures=1), DownloadInterruptionHandler(number_of_failures=3) are aimed to test the aforementioned two scenarios.

In case if malformed json/metadata are received aktualizr doesn't do any retry it just simply exits a current uptane cycle and starts a new one, so just one scenario in this case.